runtime/src/iree/modules/hal/module.c

```diff
@@ -321,6 +337,51 @@ IREE_VM_ABI_EXPORT(iree_hal_module_ex_file_from_memory,  //
 // iree_hal_allocator_t
 //===----------------------------------------------------------------------===//
 
+IREE_VM_ABI_EXPORT(iree_hal_module_allocator_select,  //
+                   iree_hal_module_state_t,           //
+                   iiICrID, rI) {
+  iree_hal_memory_type_t memory_types = (iree_hal_memory_type_t)args->i0;
+  iree_hal_buffer_usage_t buffer_usage = (iree_hal_buffer_usage_t)args->i1;
+  iree_hal_module_device_allocator_select_flags_t flags =
+      (iree_hal_module_device_allocator_select_flags_t)args->i2;
+
+  if (args->a3_count == 0) {
+    return iree_make_status(
+        IREE_STATUS_INVALID_ARGUMENT,
+        "at least one device/queue affinity pair must be provided");
+  } else if (!state->device_policy.allocator_select.fn) {
+    // No policy defined, choose the first.
+    iree_hal_device_t* first_device = NULL;
+    IREE_RETURN_IF_ERROR(
+        iree_hal_device_check_deref(args->a3[0].r0, &first_device));
+    rets->r0 = iree_hal_device_retain_ref(first_device);
+    rets->i1 = args->a3[0].i1;
+    return iree_ok_status();
+  }
+
+  iree_host_size_t pair_count = iree_hal_cast_host_size(args->a3_count);
+  if (pair_count > 64) {
+    return iree_make_status(IREE_STATUS_RESOURCE_EXHAUSTED,
+                            "too many devices for a single select call");
+  }
```

Comment:
Could include the `pair_count` value in the status message here (might be able to spot underflow or uninitialized memory better that way)